tools/src/wyvern/target/corewyvernIL/astvisitor/TailCallVisitor.java

```diff
@@ -0,0 +1,240 @@
+import wyvern.target.corewyvernIL.expression.Cast;
+import wyvern.target.corewyvernIL.expression.Expression;
+import wyvern.target.corewyvernIL.expression.FFIImport;
+import wyvern.target.corewyvernIL.expression.FieldGet;
+import wyvern.target.corewyvernIL.expression.FieldSet;
+import wyvern.target.corewyvernIL.expression.IExpr;
+import wyvern.target.corewyvernIL.expression.IntegerLiteral;
+import wyvern.target.corewyvernIL.expression.Let;
+import wyvern.target.corewyvernIL.expression.Match;
+import wyvern.target.corewyvernIL.expression.MethodCall;
+import wyvern.target.corewyvernIL.expression.New;
+import wyvern.target.corewyvernIL.expression.Path;
+import wyvern.target.corewyvernIL.expression.RationalLiteral;
+import wyvern.target.corewyvernIL.expression.StringLiteral;
+import wyvern.target.corewyvernIL.expression.Variable;
+import wyvern.target.corewyvernIL.metadata.IsTailCall;
+import wyvern.target.corewyvernIL.support.EmptyTypeContext;
+import wyvern.target.corewyvernIL.support.TypeContext;
+import wyvern.target.corewyvernIL.support.Util;
+import wyvern.target.corewyvernIL.type.CaseType;
+import wyvern.target.corewyvernIL.type.DataType;
+import wyvern.target.corewyvernIL.type.ExtensibleTagType;
+import wyvern.target.corewyvernIL.type.NominalType;
+import wyvern.target.corewyvernIL.type.StructuralType;
+import wyvern.target.corewyvernIL.type.ValueType;
+
+/** TailCallVisitor finds all MethodCall expressions in a
+ *  corewyvernIL AST and attaches an IsTailCall metadata to them
+ *  if the method calls are tail calls.
+ */
+public class TailCallVisitor extends ASTVisitor<Boolean, Void> {
+
+    public static void annotate(IExpr program) {
+        program.acceptVisitor(new TailCallVisitor(), false);
+    }
+
+    public Void visit(Boolean inTailPosition, New newExpr) {
+        for (Declaration decl : newExpr.getDecls()) {
+            decl.acceptVisitor(this, inTailPosition);
+        }
+        return null;
+    }
+
+    public Void visit(Boolean inTailPosition, Case c) {
+        c.getBody().acceptVisitor(this, inTailPosition);
+        return null;
+    }
+
+    public Void visit(Boolean inTailPosition, MethodCall methodCall) {
+        methodCall.getObjectExpr().acceptVisitor(this, false);
+        for (IExpr arg : methodCall.getArgs()) {
+            arg.acceptVisitor(this, false);
+        }
+
+        if (inTailPosition) {
+            methodCall.addMetadata(new IsTailCall());
+        }
+        return null;
+    }
+
+
+    public Void visit(Boolean inTailPosition, Match match) {
+        match.getMatchExpr().acceptVisitor(this, false);
+        match.getElseExpr().acceptVisitor(this, inTailPosition);
+        for (Case matchCase : match.getCases()) {
+            matchCase.getBody().acceptVisitor(this, inTailPosition);
+        }
+        return null;
+    }
+
+
+    public Void visit(Boolean inTailPosition, FieldGet fieldGet) {
+        fieldGet.getObjectExpr().acceptVisitor(this, inTailPosition);
```

Comment:
The recursive call should pass "false" because we have to get a field from any result